File: 프로그래머스/2/68936. 쿼드압축 후 개수 세기/쿼드압축 후 개수 세기.py
def Quad(arr, x, y, size):
    start = arr[x][y]
    same = True
    
    for i in range(x, x + size):
        for j in range(y, y + size):
            if arr[i][j] != start:
                same = False
                break
        if not same:
            break
    if same:
        return [1, 0] if start == 0 else [0, 1]
    else:
        half_size = size // 2
        top_left = Quad(arr, x, y, half_size)
        top_right = Quad(arr, x, y + half_size, half_size)
        bottom_left = Quad(arr, x + half_size, y, half_size)
        bottom_right = Quad(arr, x + half_size, y + half_size, half_size)
    return [
            top_left[0] + top_right[0] + bottom_left[0] + bottom_right[0],
            top_left[1] + top_right[1] + bottom_left[1] + bottom_right[1]
        ]
    

def solution(arr):
    answer = []
    size = len(arr)
    
    return Quad(arr, 0, 0, size)

File: 프로그래머스/2/68936. 쿼드압축 후 개수 세기/쿼드압축 후 개수 세기.java
class Solution {
    private int zero = 0;
    private int one = 0;
    
    public int[] solution(int[][] arr) {
        int n = arr.length;
        compress(arr, 0, 0, n);
        int[] answer = new int[]{zero, one};
        return answer;
    }
    
    private void compress(int[][] arr, int x, int y, int size) {
        if (isSame(arr, x, y, size)) {
            if (arr[x][y] == 0)
                zero++;
            else
                one++;
            return;
        }
        
        int half = size / 2;
        
        compress(arr, x, y, half);
        compress(arr, x, y + half, half);
        compress(arr, x + half, y, half);
        compress(arr, x + half, y + half, half);
    }
    
    private boolean isSame(int[][] arr, int x, int y, int size) {
        int num = arr[x][y];
        for (int i = x; i < x + size; i++) {
            for (int j = y; j < y + size; j++) {
                if (arr[i][j] != num)
                    return false;
            }
        }
        return true;
    }
}